doc/ExternalDependencies.md

```diff
@@ -5,56 +5,147 @@ External Dependencies
+  * [Core External Dependencies](#core-external-dependencies)
+    + [Provide Important Functionality](#provide-important-functionality)
+    + [Project Maturity](#project-maturity)
+    + [Limited Sub-dependencies](#limited-sub-dependencies)
+  * [Board-Specific External Dependencies](#board-specific-external-dependencies)
+- [Including the Dependency](#including-the-dependency)
+  * [Including Core External Dependencies](#including-core-external-dependencies)
+  * [Including Board-Specific External Dependencies](#including-board-specific-external-dependencies)
+  * [Documenting the Dependency and its Tree](#documenting-the-dependency-and-its-tree)
 
 <!-- tocstop -->
 
-## Adding a dependency
+Tock's general policy is the kernel does not include external dependencies (i.e.
+rust crates outside of the `tock/tock` repository) that are not part of the Rust
+standard library. However, on a limited, case-by-case basis with appropriate
+safeguards, external dependencies can be used in the Tock kernel. The rationale
+and policy for this is described in this document.
 
-External dependencies can be added to Tock on a case by case basis. Each
-dependency will be reviewed for inclusion. Below is a general list of
-requirements for adding a new external crate.
 
-### Provide Important Functionality
+
+## Limited External Dependency Rationale
+
+Tock limits its use of external libraries for all crates in the kernel. This is
+done to promote safety, as auditing the Tock code only requires inspecting the
+code in the Tock repository. Tock tries to be very specific with its use of
+`unsafe`, and tries to ensure that when it is used it is clear as to why. With
+external dependencies, verifying uses of `unsafe` are valid is more challenging
+to, particularly as external libraries evolve.
+
+External dependencies also typically themselves rely on dependencies, so
+including one external crate likely pulls in several external crate. As of Nov
+2022, cargo provides no mechanism for auditing and prohibiting `unsafe` in a
+dependency hierarchy. Also, the dependency chain for an external crate is
+largely hidden from developers using the external crate. Lacking automated
+tools, managing dependencies is a manual process, and to limit overhead Tock
+generally avoids external dependencies.
+
+
+## External Dependency Selection
+
+External dependencies can be added to Tock on a case-by-case basis. Each
+dependency will be reviewed for inclusion, according to the criteria in this
+section. The requirements are intentionally strict.
+
+There are two general methods to for including an external dependency in the
+Tock kernel: core external dependencies, and board-specific external
+dependencies. Core external dependencies may be used in "core" Tock crates, such
+as the `kernel`, `chips`, and `capsules` crates. Board-specific external
+dependencies may _only_ be used by crates in the `board/` folder. The processes
+for inclusion between these two methods are different.
```

Comment:
For crypto crates that will ultimately need to be used by the core kernel, this makes sense to me -- it is inevitable that those crates will need to be compiled by all boards unless we want to rely on some hacky conditional includes that will be a mess to maintain.

For external dependencies which are *only* needed by individual capsules, I think things look a little different -- we really don't want to force all boards to compile those external crates if they are not using those capsules, so I think any such capsules should exist in stand-alone crates rather than inside the `capsules` crate.

`ghash` is an interesting example because currently it is only needed in capsules, but my understanding is that the plan is to eventually use it as part of the kernel app verification workflow (is this correct, @phil-levis ?), which would require the kernel to depend on it anyway. So I don't know that we need to move Alistair's capsule into a standalone crate.

Reply:
What's the right place to put the separation? I worry that we could be including fine-grained slicing before it's clear we need to. Does anyone care if the first build takes a couple seconds longer? I agree that once it's minutes then that is a lot of overhead, but perhaps we cross that bridge when we get there.

We also could look at separating capsules in to multiple crates, which could address this issue as well. Or put all capsules-with-dependencies in their own crate.

Reply:
`ghash` isn't specifically for app signing. The idea with the `ghash` crate is that we can expose AES-GCM on all platforms that have any AES hardware. The goal here would be to support all AES block ciphers (GCM, CCM, CTR, CBC and ECB) on any platform that supports AES. That way we can rely on all implementations existing and expose them all to userspace and the kernel.

Basically CTR, CBC and ECB are commonly supported in hardware, and the other two can be built with software on top of the hardware.

In the longer run though, I think it would be worth supporting crypto operations (AES, RSA, ECC) on all platforms with software fallbacks. The goal there is to allow loading signed/encrypted applications on all Tock platforms. As a secure operating system, it seems very useful for Tock to be able to do that.

I do agree with Brad. It does seem like we are worrying about something that isn't a problem yet. In terms of crypto crates we only ever expect to include a handful. We can also do the `external-capsules` direction, but I feel like the crypto ones might end up being included all the time anyway

Reply:
Yeah, `ghash` isn't needed for application credentials. However, public key signatures (RSA, ECDSA, DH, etc.) will need to use tested-and-true software implementations (existing crates).